crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -1266,18 +1274,126 @@ pub(crate) mod testing {
+    ///         }
+    ///         "usage": [ "master" ],
+    ///         "user_id": "their_user_id"
+    ///     }
+    /// });
+    ///
+    /// let ssk_json = json!({
+    ///     "their_user_id": {
+    ///         "keys": { "ed25519:blah": "blah" },
+    ///         "signatures": {
+    ///             "their_user_id": { "ed25519:blah": "blah" }
+    ///         },
+    ///         "usage": [ "self_signing" ],
+    ///         "user_id": "their_user_id"
+    ///     }
+    /// })
+    ///
+    /// let response = simulate_key_query_response_for_verification(
+    ///     signature_upload_request,
+    ///     my_identity,
+    ///     my_user_id,
+    ///     their_user_id,
+    ///     msk_json,
+    ///     ssk_json
+    /// ).await;
+    ///
+    /// olm_machine
+    ///     .mark_request_as_sent(
+    ///         &TransactionId::new(),
+    ///         crate::IncomingResponse::KeysQuery(&kq_response),
+    ///     )
+    ///     .await
+    ///     .unwrap();
+    /// ```
+    pub fn simulate_key_query_response_for_verification(
+        signature_upload_request: SignatureUploadRequest,
+        my_identity: OwnUserIdentity,
+        my_user_id: &UserId,
+        their_user_id: &UserId,
+        msk_json: serde_json::Value,
+        ssk_json: serde_json::Value,
+    ) -> KeyQueryResponse {
+        // Find the signed key inside the SignatureUploadRequest
+        let cross_signing_key: CrossSigningKey = serde_json::from_str(
+            signature_upload_request
+                .signed_keys
+                .get(their_user_id)
+                .expect("Signature upload request should contain a key for their user ID")
+                .iter()
+                .next()
+                .expect("There should be a key in the signature upload request")
+                .1
+                .get(),
+        )
+        .expect("Should not fail to deserialize the key");
+
+        // Find their master key that we want to update inside their msk JSON
+        let mut their_msk: CrossSigningKey = serde_json::from_value(
+            msk_json.get(their_user_id.as_str()).expect("msk should contain their user ID").clone(),
+        )
+        .expect("Should not fail to deserialize msk");
+
+        // Find our own user signing key
+        let my_user_signing_key_id = my_identity
+            .user_signing_key()
+            .keys()
+            .iter()
+            .next()
+            .expect("There should be a user signing key")
+            .0;
+
+        // Add the signature from the SignatureUploadRequest to their master key, under
+        // our user ID
+        their_msk.signatures.add_signature(
+            my_user_id.to_owned(),
+            my_user_signing_key_id.to_owned(),
+            cross_signing_key
+                .signatures
+                .get_signature(my_user_id, my_user_signing_key_id)
+                .expect("There should be a signature for our user"),
```

Comment:
suggest moving this up to next to where we construct `cross_signing_key`. It's only needed for this one operation and conceptually they fit together.